### PR TITLE
Remove duplicate GC routes on summarization

### LINKED ISSUE #9472: GC data is duplicated and is very large in table heavy files
Fluid DocId == “aMtUNIYUVCTh9pnXng3MeXBc4k11eKpqrj4xeLK8S3A%3D”
[Actual File](https://www.office.com/launch/fluid/content?auth=2&action=edit&drive=b!sT6xOCLunUaw96VVvQpfKDMINLoURjxDlZOHWW-VJzmXkPSR7oYpSYFq_u8nsw1D&item=01H3KDHNH6PY5TBAJNXFD3MQGLPCNQTUPZ&file=https:%2F%2Fmicrosoft-my.sharepoint.com%2Fpersonal%2Fdevverma_microsoft_com&siteUrl=https:%2F%2Fmicrosoft-my.sharepoint.com%2Fpersonal%2Fdevverma_microsoft_com)

Data_eventName | Data_deletedNodes | Data_totalNodes | Data_deletedDataStores | Data_totalDataStores | count_
-- | -- | -- | -- | -- | --
fluid:telemetry:Summarizer:Running:GarbageCollection_end | 599 | 76,566 | 162 | 1,302 | 7

GC blobs are 44Mb in size (1378 unique blobs)!
So I think we have to take a closer look at GC in general, and telemetry to catch issues like that (more on one angle below)
 
There is a lot of replicated data that feels like should not be there in GC blobs. I.e.:
{
  "gcData": {
    "gcNodes": {
      "/d016fa27-8dfc-4d14-b41a-db660dede561": [
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae"
      ],
      "/72436162-1b8b-4cc6-a155-b77d13beeca7": [
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae",
        "/571dd49a-6477-4807-99d9-5808f70ddcae"
      ],

### PR DESCRIPTION
**Pull request by tyler-cai-microsoft**
Resolves #9472 

We've discovered multiple duplicate routes in our GC outbound routes in our GC blob. This PR's goal is to prevent that from happening and will fix the bug on summarization. 

How to replicate the issue:
1. Have a datastore with 2+ DDSs. 
2. Run GC's `collectGarbage` during summarization, which the parent datastore adds GC routes from each DDS to itself during the `getGCData` call.
3. The summary is submitted to the server.
4. Reload a new container, which has these newly added routes. This is then passed down from the container runtime to the datastore back down to the child datastores' `summarizerNodeWithGC` as `this.gcData`. 
5. Make a change to one of the DDSs but not another.
6. Re-run GC's `collectGarbage' during summarization
7. The DDS which did not change returns `getGCData` with the data passed down from the snapshot containing an extra route.
8. The parent datastore adds the GC route from that DDS to itself again, which is a duplicate route.
9. The summary is submitted to the server.

Solution:
- When a DDS's `gcData` is passed back down to it, remove the implicitly added route to its parent datastore.

This relies on Snapshot tests being setup #10015

### CODE REVIEW

**Comment by vladsud** [comment]
Any chance you can describe in plain English what is the problem, and how big is it? I glanced through changes, but it's hard for me to get a gist of it quickly :)
I.e., does it get worse over time? How many duplicate entries we get on average? How big impact do you think it will have in terms of size?

**Comment by tyler-cai-microsoft** [comment]
> Any chance you can describe in plain English what is the problem, and how big is it? I glanced through changes, but it's hard for me to get a gist of it quickly :) I.e., does it get worse over time? How many duplicate entries we get on average? How big impact do you think it will have in terms of size?

This means the GC blob will get bigger overtime. We don't have data yet on the size of GC blobs until 0.59.1000 is integrated. This issue affects all components/datastores with multiple DDSs.

I've updated the comment to include more details. I'll have to get back to you on duplicate entries on average and impact in terms of size.

**Comment by vladsud** [comment]
Do we need to do anything to "reset" existing files? 
- Will re-summarization of data store fix that data store representation?
- Do we need to reset GC version to force full summary (or use any other mechanism to recalc this data)?

**Comment by tyler-cai-microsoft** [comment]
> Do we need to do anything to "reset" existing files?
> 
> * Will re-summarization of data store fix that data store representation?
> * Do we need to reset GC version to force full summary (or use any other mechanism to recalc this data)?

To reset existing files, we need to re-summarize all DDSs with this issue. 

I just realized that my current solution does not fix old summaries as it only deletes one route. We can implement a solution that deletes all routes as soon as summarization occurs. There is one edge case we have to consider - if a DDS explicitly adds a route to its parent, then the gcData we would be passing down to the datastore would not be important. The route would automatically be added back. In the future if we make DDSs non-root this will likely need to be reconsidered/remembered.

**Comment by vladsud** [comment]
> if a DDS explicitly adds a route to its parent

DDS should always have a link back to data store - if DDS is alive (because someone holds it), then Data store should be alive.

Root DDS means Data store has a reference to it - DDS would not be collected if data store is alive.
Non-root means Data store does not keep it alive - if there are no other references, it should be collected (whenever / if we implement GC for DDSs).

**Comment by agarwal-navin** [comment]
The simplest solution is to use a set instead of an array while processing outbound routes in either direction - sending data up from DDS to container runtime or sending it down from container runtime to DDS. Then explicitly remove the link from the DDS to the data store. This will ensure that there are never any duplicate routes in the GC data.

> if a DDS explicitly adds a route to its parent, then the gcData we would be passing down to the datastore would not be correct.

From GC's perspective, it doesn't matter whether a route from DDS to its parent data store was explicitly added by the DDS or implicit added to keep the data store alive. Another thing to consider is if a DDS has multiple handles to the same object, it will have a lot of duplicate outbound routes. GC should remove these duplicates as well before it processes them and stores them in the summary.

> In the future if we make DDSs non-root this will likely need to be reconsidered/remembered.

I am not sure I understand this. DDSs being root or non-root has nothing to do with the route back to the data store. That is about having a route from the data store to its child DDSs.

**Comment by vladsud** [comment]
I'd only say - we need to stop the bleeding and fix existing files. If this can be done in one go - fine. If not - consider targeted fix to stop the bleeding (and possibly porting / producing hotfixes).

This is also an area where we need to rethink validation. Ideally, we should find them sooner and via some automated ways. We should consider multiple directions - better telemetry (upfront), better UTs (we should have UTs that keep summarizing 100 times with various changes or not to container and observe metrics), etc.

**Comment by tyler-cai-microsoft** [comment]
Given the size of the average GC blob ~ 29 within the last day and an estimation of about 29 routes per blob and an estimation of 76 bytes per route, which is the length of 2 full gc id routes with "/" delimiters, the average expected GC blob size is estimated to be 63916 bytes or ~64 KB.

**Comment by vladsud** [comment]
Looks reasonable to me, but Navin should be the one signing off :)
Is this enough to fix everything, even if some number of data stores would never summarize in a given file (user would not change them)?

**Comment by agarwal-navin** [comment]
> > Looks reasonable to me, but Navin should be the one signing off :) Is this enough to fix everything, even if some number of data stores would never summarize in a given file (user would not change them)?
> 
> Yes, because the `GCDataBuilder` is used during the `getGCData` call at the container runtime level, so at that point, any duplicate routes would have been removed - including for duplicate routes we do not already know about.

I believe so. You can validate this with a test to confirm. Also, the snapshot tests which you are working on will also validate this.

File: packages/runtime/datastore/src/dataStoreRuntime.ts [hunk 1]
```diff
@@ -675,6 +675,15 @@ IFluidDataStoreChannel, IFluidDataStoreRuntime, IFluidHandleContext {
         let channelBaseGCDetails = (await this.channelsBaseGCDetails).get(channelId);
         if (channelBaseGCDetails === undefined) {
             channelBaseGCDetails = {};
+        } else if (channelBaseGCDetails.gcData?.gcNodes !== undefined) {
```
Comment (agarwal-navin) [inline packages/runtime/datastore/src/dataStoreRuntime.ts:678]: I just realized that this bug was a regression added by this PR - https://github.com/microsoft/FluidFramework/pull/8389. Take a look at `this.initialChannelGCDataP` which was removed / updated by the above PR from `dataStoreRuntime.ts`. It called `removeRouteFromAllNodes(gcData.gcNodes, this.absolutePath);` to remove the back routes which was removed :-(

Comment (agarwal-navin) [reply packages/runtime/datastore/src/dataStoreRuntime.ts:678]: nit: You could also do what was happening before this bug was introduced, i.e., call `removeRouteFromAllNodes(gcData.gcNodes, this.absolutePath)` in `this.channelsBaseGCDetails` after getting the baseGCDetails which is initialized in the data store runtime constructor. That should be a one line change.

File: packages/runtime/garbage-collector/src/utils.ts [hunk 2]
```diff
@@ -179,10 +179,17 @@ export function concatGarbageCollectionData(gcData1: IGarbageCollectionData, gcD
...
+        const gcNodes = {};
+        for (const [nodeId, outboundRoutes] of Object.entries(this.gcNodesSet)) {
+            gcNodes[nodeId] = [...outboundRoutes];
+        }
+        return gcNodes;
+    }
 
     public addNode(id: string, outboundRoutes: string[]) {
-        this.gcNodes[id] = Array.from(outboundRoutes);
+        this.gcNodesSet[id] = new Set(outboundRoutes);
```
Comment (agarwal-navin) [inline packages/runtime/garbage-collector/src/utils.ts:192]: `unpackChildNodesGCDetails` should also be updated to form a Set from the outboundRoutes. That is the function that is used when sending the GC data down from container runtime to DDSs. That way you don't have to individually convert them to set at each layer.
It will also remove all duplicate routes across container runtime -> data store boundary.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+import { ISummaryBlob, ISummaryTree } from "@fluidframework/protocol-definitions";
+import { SharedMap } from "@fluidframework/map";
+import { ISummaryContext } from "@fluidframework/driver-definitions";
+import { TelemetryNullLogger } from "@fluidframework/common-utils";
+import { TestDataObject, loadSummarizer, submitAndAckSummary } from "../mockSummarizerClient";
+import { mockConfigProvider } from "./mockConfigProivder";
+import { wrapDocumentServiceFactory } from "./gcDriverWrappers";
+
+/**
+ * Validates this scenario: When a datastore is aliased that it is considered a root datastore and always referenced
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:32]: This comment needs to be updated.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+        [innerRequestHandler],
+        runtimeOptions,
+    );
+
+    const settings = {
+        "Fluid.GarbageCollection.LogUnknownOutboundRoutes": "true",
+        "Fluid.GarbageCollection.WriteDataAtRoot": "true",
+    };
+    const configProvider = mockConfigProvider(settings);
+    settings["Fluid.GarbageCollection.UseDataStoreAliasing"] = "false";
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:67]: This can be moved to where settings is initialized.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+        latestAckedSummary = summaryResult.ackedSummary;
+        assert(latestUploadedSummary !== undefined, "Did not get a summary");
+        return getGCContentFromTree(latestUploadedSummary);
+    }
+
+    const createContainer = async (): Promise<IContainer> => {
+        return provider.createContainer(runtimeFactory, { configProvider });
+    };
+
+    const getGCContentFromTree = (summary: ISummaryTree): IGarbageCollectionState => {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:113]: nit: This can be inlined in `summarizeOnNewContainerAndGetGCTree` since there is only one call.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+
+        // Change ds1 but not the root dds
+        dds.set("change", "change1");
+
+        assert(latestAckedSummary !== undefined, "Ack'd summary isn't available as expected");
+        const gcTree = await summarizeOnNewContainerAndGetGCTree(latestAckedSummary.summaryAck.contents.handle);
+        assert(gcTree !== undefined, "Expected a gc tree!");
+        const routes = gcTree.gcNodes[`/${mainDataStore.id}/${mainDataStore._root.id}`].outboundRoutes;
+        const seenRoutes = new Set<string>();
+        routes.forEach((route) => {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:149]: It may be worth checking that none of the nodes have duplicate routes. Maybe a helper function that does this. Eventually we can make that part of all our GC tests that validate the GC data. That will give us coverage in more scenarios.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts (hunk 4 above)
Comment (tyler-cai-microsoft) [reply packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:67]: DataStoreAliasing is set to true by default when we create the mockConfigProvider, so we have to set it to false afterwards. I don't think we care about this case anyway as this issue happens at the DDS level more than the DS level.

File: packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        if (latestAckedSummary !== undefined) {
+            newSummaryContext.ackHandle = latestAckedSummary.summaryAck.contents.handle;
+            newSummaryContext.proposalHandle = latestAckedSummary.summaryOp.contents.handle;
+        }
+        return newSummaryContext;
+    }
+
+    const logger = new TelemetryNullLogger();
+
+    async function summarizeOnNewContainerAndGetGCTree(summaryVersion?: string): Promise<IGarbageCollectionState> {
```
Comment (agarwal-navin) [inline packages/test/test-end-to-end-tests/src/test/gc/gcDatastoreDuplicateRoutes.spec.ts:94]: nit: `summarizeOnNewContainerAndGetGCState`

**Review by agarwal-navin — APPROVED** [review]

**Comment by agarwal-navin** [comment]
I don't think you can update the test snapshots in this PR because it fixes the bug. Maybe you can if you update the snapshots in your local repo with this change. This will add a "src_snapshots" folder (which has duplicate routes) to the snapshots added by https://github.com/microsoft/FluidFrameworkTestData/pull/41 and update the "current_snapshots" to not have the duplicate routes.

**Comment by agarwal-navin** [comment]
Also, please make sure that the snapshot repo changes are merged and you update this PR to point to a commit in the main branch of the snapshot repo.
